### PR TITLE
Add processor for refuse a challenge, check for state and permission …

### PR DESCRIPTION
**Pull request by rydogs**
…in callback processor

### CODE REVIEW

File: src/main/java/com/offdk/play/model/game/MatchPlayer.java [hunk 1]
```diff
@@ -7,28 +7,29 @@
...
-  MatchPlayer(String id, MatchPlayerType type, Optional<Integer> score, Optional<BigDecimal> rating) {
-    this.id = id;
+  MatchPlayer(User user, MatchPlayerType type, Optional<Integer> score, Optional<BigDecimal> rating) {
+    this.user = user;
     this.type = type;
     this.score = score;
     this.rating = rating;
   }
 
   @VisibleForTesting
```
Comment (dfurstein) [inline src/main/java/com/offdk/play/model/game/MatchPlayer.java:30]: Looks like you left this behind mainly cause you didn't want the test to fail, but I had the rating living on the Player object. If the MatchPlayer object has no relationship to that object, how does it have/get a rating?

Comment (rydogs) [reply src/main/java/com/offdk/play/model/game/MatchPlayer.java:30]: No, this was needed.  MatchPlayer object is a snapshot of an player (drumroll... just like identity).  A matchPlayer is carries additional information like score and player type (challenger vs challenged).  Rating calculator returns a list of rating, which can be set on player objects separately.

To set player rating, I would add a method call `updatePlayerRating(String id, BigDecimal newRating)` in PlayerRep which will set the player rating after the game using data from MatchPlayer.

As much as I would like to use player for everything, it just doesn't make sense given what we have to do.

Also, i would like to discuss if we really need to keep multiple ratings per player.  I don't think this is necessary as you will never compare multiple ratings at any given time.  If we decide to switch rating system, i would rather recalculate.

File: src/main/java/com/offdk/play/model/game/MatchStatus.java [hunk 2]
```diff
@@ -3,7 +3,7 @@
 public enum MatchStatus {
   CHALLENGED,
   ACCEPTED,
-  REJECTED,
+  REFUSED,
```
Comment (dfurstein) [inline src/main/java/com/offdk/play/model/game/MatchStatus.java:6]: Why the change?

File: src/main/java/com/offdk/play/service/RefuseMatchCallbackPrecessor.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  @Override
+  public Message process(CallbackRequest request) {
+    return Try.of(() -> matchRepo.findOne(request.callbackId()))
+        .filter(isChallegedUser(request.user()))
+        .onFailure(MatchException.ofThrow("Only the challenged can refuse the game"))
+        .filter(challengedStatus())
+        .map(m -> {
+          m.accept();
```
Comment (dfurstein) [inline src/main/java/com/offdk/play/service/RefuseMatchCallbackPrecessor.java]: Copy and paste is not your friend.

File: src/main/java/com/offdk/play/model/game/MatchPlayer.java (hunk 1 above)
Comment (dfurstein) [reply src/main/java/com/offdk/play/model/game/MatchPlayer.java:30]: Aren't we using mongo? Is it really that difficult to store each rating as it's own data field within a map? Even if we don't add a second rating system, I could easily see a need to have an overall rating and a seasonal rating, at the same time.

**Review by dfurstein — DISMISSED** [review]
RefuseMatchCallbackPrecessor has approval logic.

File: src/main/java/com/offdk/play/model/game/MatchPlayer.java (hunk 1 above)
Comment (rydogs) [reply src/main/java/com/offdk/play/model/game/MatchPlayer.java:30]: It is not the fact that it is difficult to store.  It is the fact that match player has mutable attributes that doesn't apply to player

File: src/main/java/com/offdk/play/model/game/MatchStatus.java (hunk 2 above)
Comment (rydogs) [reply src/main/java/com/offdk/play/model/game/MatchStatus.java:6]: To make things more consistent.  We had refuse and reject in interchangeably.  I am fine with either one as long as we use the same word.  The buttons currently have refuse.

**Comment by rydogs** [comment]
Fixed logic in RefuseMatchCallbackProcessor

File: src/main/java/com/offdk/play/model/game/MatchPlayer.java (hunk 1 above)
Comment (dfurstein) [reply src/main/java/com/offdk/play/model/game/MatchPlayer.java:30]: Not sure how allowing multiple ratings would impact MatchPlayer, if you want the ratings to live on Player.

But if you do want this break, no reason to leave this constructor around anymore. It's just cluttering the code and the test cases should probably be fixed to reflect that change.

File: src/main/java/com/offdk/play/service/RefuseMatchCallbackPrecessor.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  public Message process(CallbackRequest request) {
+    return Try.of(() -> matchRepo.findOne(request.callbackId()))
+        .filter(isChallegedUser(request.user()))
+        .onFailure(MatchException.ofThrow("Only the challenged can refuse the game"))
+        .filter(challengedStatus())
+        .map(m -> {
+          m.refuse();
+          matchRepo.save(m);
+          return Message.createSuccess(m.refuseMatchText()).build();
+        }).getOrElseThrow(() -> MatchException.of("Match has already started"));
```
Comment (dfurstein) [inline src/main/java/com/offdk/play/service/RefuseMatchCallbackPrecessor.java]: If the match has been refused, it cannot have already started 👅

**Review by dfurstein — APPROVED** [review]
Make that one change as discussed, and merge whenever
